// solr-faceted-search-react/src/components/search-field-container.js
import PropTypes from 'prop-types';
import React from "react";
import cx from "classnames";
import DjangoCSRFToken from "../django-csrf-token"
import solrQuery2 from "../api/solr-query";

class PostQueryButton extends React.Component {

	render() {

		const { bootstrapCss, query, numFound } = this.props;

		let root = document.getElementById('app');
		let cible = root.getAttribute('cible');
		//onSubmit={(e) => {this.handleSubmit(e)}}
		return (
			<form action={"query/" + cible + "/"} method="POST" >
				<DjangoCSRFToken />
				{/* 
				First input is the query in SolrClient raw format
				It's ready to be used by this very specific client
			*/}
				<input
					type="hidden"
					name="solr_params"
					value={query ? JSON.stringify(query) : '{"error":"no_query"}'}
				/>
				{/* 
				Second input is the query in solr json format
				It's ready to be used by any other solr client
			*/}
				<input
					type="hidden"
					name="solr_json_query"
					value={query ? solrQuery2(query) : '{"error":"no_query"}'}
				/>
				<input type="hidden" name="count" value={numFound} />
				<button id="btn-submit" className={cx({ "btn": bootstrapCss, "btn-default": bootstrapCss, "btn-sm": bootstrapCss, "pull-right": bootstrapCss, "btn-success": bootstrapCss })}
					type='submit'>
					Valider ma sélection
	 		</button>
			</form>)
	}
}

PostQueryButton.propTypes = {
	bootstrapCss: PropTypes.bool,
	onChange: PropTypes.func,
	query: PropTypes.object,
	numFound: PropTypes.number
};

class SearchFieldContainer extends React.Component {

	render() {
		const { bootstrapCss, onNewSearch, query, onChangeQuery, numFound } = this.props;

		return (
			<div className={cx({ "col-md-9 col-sm-8": bootstrapCss })}>
				<div id="current-query-div" className={cx({ "panel": bootstrapCss, "panel-default": bootstrapCss })}>
					<header className={cx({ "panel-heading": bootstrapCss })}>
						<PostQueryButton query={query} onChange={onChangeQuery} bootstrapCss={bootstrapCss} numFound={numFound} />
						<button className={cx({ "btn": bootstrapCss, "btn-default": bootstrapCss, "btn-sm": bootstrapCss, "pull-right": bootstrapCss })}
							onClick={onNewSearch}>
							Réinitialiser
						</button>
						<label id="search-label">Recherche</label>
					</header>
					{this.props.children[0]}
					<ul className={cx("solr-search-fields", { "list-group": bootstrapCss })}>
						{this.props.children.slice(1)}
					</ul>
					<PostQueryButton query={query} onChange={onChangeQuery} bootstrapCss={bootstrapCss} numFound={numFound} />
					<button className={cx({ "btn": bootstrapCss, "btn-default": bootstrapCss, "btn-sm": bootstrapCss, "pull-right": bootstrapCss })}
						onClick={onNewSearch}>
						Réinitialiser
					</button>
				</div>
			</div>
		);
	}
}

SearchFieldContainer.propTypes = {
	bootstrapCss: PropTypes.bool,
	children: PropTypes.array,
	onNewSearch: PropTypes.func,
	query: PropTypes.object
};

export default SearchFieldContainer;
