// solr-faceted-search-react/src/components/matrix-facet/index.js
import PropTypes from 'prop-types';
import React from "react";
import cx from "classnames";

import CheckedIcon from "../icons/checked";
import UncheckedIcon from "../icons/unchecked";

class MatrixCell extends React.Component {

	// toggleCell(direction,current_cell_status) {
	// 	var new_cell_status = current_cell_status;
	// 	switch(direction){
	// 		case -1:
	// 			if (current_cell_status != 0) {new_cell_status=0;} else {new_cell_status= -1;}
	// 			break;
	// 		case 1:
	// 			if (current_cell_status != 0) {new_cell_status=0;} else {new_cell_status= 1;}
	// 			break;
	// 		default:break;}
	// 		return new_cell_status;
	// }
	//
	// handleClick(e,value){
	// 			var newCellStatus = 0;
	// 			if(e.shiftKey) {
	// 				/* Shift Click excludes or reset */
	// 				 newCellStatus = this.toggleCell(-1,this.props.cell.status);
	// 			} else {
	// 				/* Click includes or reset */
	// 				 newCellStatus = this.toggleCell(1,this.props.cell.status);
	// 			}
	//
	// 			const cell = [this.props.row_id,this.props.col_id,value,newCellStatus].join('.');
  //   		console.log('Higway to cell '+cell);
	// 			const foundIdx = this.props.value.indexOf(cell);
	// 			if (foundIdx < 0) {
	// 				this.props.onChange(this.props.field, this.props.value.concat(cell));
	// 			} else {
	// 				this.props.onChange(this.props.field, this.props.value.filter((v, i) => i !== foundIdx));
	// 			}
	// }

	render () {
		let bootstrapCss = true;

		return (
			<button className = {cx({"btn btn-default": bootstrapCss},
									{'selected': this.props.cell.status > 0},
									{'rejected': this.props.cell.status < 0})}
					onClick={(e) => this.props.cellClickHandler(e,this.props.cell.camp,
																this.props.row_id,
																this.props.col_id,
																this.props.cell.status)}>
				{this.props.cell.camp}
			</button>
		)
	}
}


class MatrixRow extends React.Component {

	render () {

		let columnHeaderStyle = {textAlign: "center"};
		let bootstrapCss = true;

		return (
			<tr className={cx('matrix-facet-table')}>
				<td styleName={columnHeaderStyle}>
					<button className={cx({"btn btn-sm": bootstrapCss},this.props.row_content.class)}>{this.props.row_name}</button>
				</td>
				<td>
					<div className = {cx({"btn-group btn-group-sm btn-block": true})}>
					{this.props.row_content.CDE.map((k) => {
						return <MatrixCell cell={k}
											row_id={this.props.row_id}
											col_id={'CDE'}
											onChange={this.props.onChange}
											field={this.props.field}
											value={this.props.value}
											cellClickHandler={this.props.cellClickHandler}/>})}
					</div>
				</td>
				<td>
					<div className = {cx({"btn-group btn-group-sm btn-block": true})}>
					{this.props.row_content.KIT.map((k)=> {
						return <MatrixCell cell={k}
											row_id={this.props.row_id}
											col_id={'KIT'}
											onChange={this.props.onChange}
											field={this.props.field}
											value={this.props.value}
											cellClickHandler={this.props.cellClickHandler}/>})}
					</div>
				</td>
			</tr>
		)
	}
}

class MatrixFacet extends React.Component {

	constructor(props) {
		super(props);

		this.state = {
			filter: "",
			truncateFacetListsAt: props.truncateFacetListsAt,
			disabled: true
		};
	}

	toggleCellStatus(current_cell_status) {
		var new_cell_status = current_cell_status;
		if (current_cell_status != 0) {
			new_cell_status=0;
		} else {
			new_cell_status= 1;}

			return new_cell_status;
	}

	handleClick(e,value,row_id,col_id,status){
				var newCellStatus = 0;
				if(e.shiftKey) {
					/* Shift Click excludes */
					 newCellStatus = -1;
				} else {
					/* Click includes or reset according to status */
					 newCellStatus = this.toggleCellStatus(status);
				}

				const cell = [row_id,col_id,value,newCellStatus].join('.');
				const shortCell = [row_id,col_id,value].join('.');
				const suffix = '.0';
				// Remove all existing values for the cell and all neutral values
				this.props.onChange(this.props.field,
									this.props.value.filter((v) => (v.indexOf(shortCell)<0 || v.indexOf(suffix, v.length - suffix.length)>0))
													.concat(cell));
	}


	toggleExpand() {
		this.props.onSetCollapse(this.props.field, !(this.props.collapse || false));
	}

	mergeMatrix(m) {
		const values = this.props.value || ['....'];
		values.map((v) => { 
			const [row,col,cell,state] = v.split('.');
			(row.length>0 && col.length) > 0 ? m[row][col].filter((c) => {
				return c.camp==cell;}).map((c) => c.status = parseInt(state)) : 0
		});
		console.log('>>>>>>>>>> MERGEMATRIX');
		console.log(m);

	}

	render() {
		const { query, label, facets, field, value, bootstrapCss, facetSort, collapse } = this.props;
		const { truncateFacetListsAt } = this.state;

		// var matrix = {
		// 	'SAV':{
		// 		'CDE':[{camp:'X',status:0},{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0}],
		// 		'KIT':[{camp:'X',status:0},{camp:'Y',status:1},{camp:'Z',status:1},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0}],
		// 		'class':'btn-danger',
		// 		'label':'Saveurs'
		// 	},
		// 	'EDF':{
		// 		'CDE':[{camp:'X',status:-1},{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0}],
		// 		'KIT':[{camp:'X',status:0},{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0}],
		// 		'class':'btn-success',
		// 		'label':'Fleurs'
		// 	},
		// 	'CRE':{
		// 		'CDE':[{camp:'X',status:0},{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0}],
		// 		'KIT':[{camp:'X',status:0},{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0}],
		// 		'class':'btn-warning',
		// 		'label':'Création'
		// 	},
		// 	'INI':{
		// 		'CDE':[{camp:'X',status:0},{camp:'Y',status:1},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0}],
		// 		'KIT':[{camp:'X',status:0},{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0}],
		// 		'class':'',
		// 		'label':'Initiatives'
		// 	}
		// }

		const matrix = {
			'SAV':{
				'CDE':[{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0},{camp:'G',status:0}],
				'KIT':[{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0},{camp:'G',status:0}],
				'class':'btn-danger',
				'label':'Saveurs'
			},
			'EDF':{
				'CDE':[{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0},{camp:'G',status:0}],
				'KIT':[{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0},{camp:'G',status:0}],
				'class':'btn-success',
				'label':'Fleurs'
			},
			'CRE':{
				'CDE':[{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0},{camp:'G',status:0}],
				'KIT':[{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0},{camp:'G',status:0}],
				'class':'btn-warning',
				'label':'Création'
			},
			'INI':{
				'CDE':[{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0},{camp:'G',status:0}],
				'KIT':[{camp:'Y',status:0},{camp:'Z',status:0},{camp:'A',status:0},{camp:'B',status:0},{camp:'F',status:0},{camp:'G',status:0}],
				'class':'btn-primary',
				'label':'Initiatives'
			}
		}

		this.mergeMatrix(matrix);

		const facetCounts = facets.buckets ? facets.buckets.map((b) => b.count) : [];
		const facetValues = facets.buckets ? facets.buckets.map((b) => b.val) : [];

		const facetSortValue = facetSort ? facetSort :
			query.facetSort ? query.facetSort :
			(query.facetLimit && query.facetLimit > -1 ? "count" : "index");

		const expanded = !(collapse || false);

		const showMoreLink = truncateFacetListsAt > -1 && truncateFacetListsAt < facetValues.length ?
			<li className={cx({"list-group-item": bootstrapCss})} onClick={() => this.setState({truncateFacetListsAt: -1})}>
				Show all ({facetValues.length})
			</li> : null;

		let columnHeaderStyle = {textAlign: "center"};

		return (
			<li className={cx("facet-box","list-facet", {"list-group-item col-md-8": bootstrapCss})} id={`solr-list-facet-${field}`}>
				<header onClick={this.toggleExpand.bind(this)}>
					<h5>
						{bootstrapCss ? (<span>
							<span className={cx("glyphicon", {
								"glyphicon-collapse-down": expanded,
								"glyphicon-collapse-up": !expanded
							})} />{" "}
						</span>) : null }
						{label}
					</h5>
				</header>
				{ expanded ? (
					<table>
						<thead>
							<tr className={cx('matrix-facet-table')}>
								<th style={columnHeaderStyle}>
									<button disabled className={cx('matrix-column',{"btn btn-info btn-sm btn-block": bootstrapCss})}>Commandes</button>
								</th>
								<th style={columnHeaderStyle}>
									<button disabled className={cx('matrix-column',{"btn btn-info btn-sm btn-block": bootstrapCss})}>Kits</button>
								</th>
							</tr>
						</thead>
						<tbody>
							{Object.keys(matrix)
									.map((k)=> {return <MatrixRow row_name={matrix[k].label}
																	row_content={matrix[k]}
																	row_id={k}
																	onChange={this.props.onChange}
																	field={this.props.field}
																	value={this.props.value}
																	cellClickHandler={this.handleClick.bind(this)}/>}
									)
							}
						</tbody>
					</table>
				) : null }
			</li>
		);
	}
}

MatrixFacet.defaultProps = {
	value: []
};

MatrixFacet.propTypes = {
	bootstrapCss: PropTypes.bool,
	children: PropTypes.array,
	collapse: PropTypes.bool,
	facetSort: PropTypes.string,
	facets: PropTypes.array.isRequired,
	field: PropTypes.string.isRequired,
	label: PropTypes.string,
	onChange: PropTypes.func,
	onFacetSortChange: PropTypes.func,
	onSetCollapse: PropTypes.func,
	query: PropTypes.object,
	truncateFacetListsAt: PropTypes.number,
	value: PropTypes.array
};

export default MatrixFacet;
