// solr-faceted-search-react/src/components/results/count-label.js
import PropTypes from 'prop-types';
import React from "react";
import cx from "classnames";

const resultCountLabels = {
	pl: "Found % results",
	sg: "Found % result",
	none: "No results"
};

class Result extends React.Component {
	render() {
		const { numFound } = this.props;
		const { bootstrapCss } = this.props;
		const resultLabel = numFound > 1 ? resultCountLabels.pl :
			numFound === 1 ? resultCountLabels.sg :
			resultCountLabels.none;

		return (
			//<label>
			//	{resultLabel.replace("%", numFound.toLocaleString().replace(',','&thinsp;'))}
			//</label>
			<div className={cx({"clearfix": bootstrapCss})}>
			<div className={cx({"row": bootstrapCss})}>
				<div className={cx({"col-md-3": bootstrapCss,"media-icon": bootstrapCss})}>
					<i className={cx("fa","fa-drivers-license-o fa-2x")} aria-hidden="true"></i>
				</div>
				<div className={cx({"col-md-6": bootstrapCss,"media-count": bootstrapCss})}>
					{numFound.toLocaleString().replace(',','&thinsp;')}
				</div>
				<div className={cx({"col-md-3": bootstrapCss,"media-target": bootstrapCss})}>
						<button type="button" className={cx({"btn": bootstrapCss,"btn-outline-primary": bootstrapCss,"btn-xs": bootstrapCss})}>Cibler</button>
				</div>
			</div>

			<div className={cx({"row": bootstrapCss})}>
				<div className={cx({"col-md-3": bootstrapCss,"media-icon": bootstrapCss,"text-center": bootstrapCss})}>
					<i className={cx("fa","fa-phone fa-2x")} aria-hidden="true"></i>
				</div>
				<div className={cx({"col-md-6": bootstrapCss,"media-count": bootstrapCss})}>10 715</div>
				<div className={cx({"col-md-3": bootstrapCss,"media-target": bootstrapCss})}>
					<button type="button" className={cx({"btn": bootstrapCss,"btn-xs": bootstrapCss})}>Cibler</button></div>
			</div>

			<div className={cx({"row": bootstrapCss})}>
				<div className={cx({"col-md-3": bootstrapCss},"media-icon": bootstrapCss,"text-center": bootstrapCss)}>
					<i className={cx("fa","fa-mobile fa-2x")} aria-hidden="true"></i>
				</div>
				<div className={cx({"col-md-6": bootstrapCss,"media-count": bootstrapCss})}>2 240</div>
				<div className={cx({"col-md-3": bootstrapCss,"media-target": bootstrapCss})}>
					<button type="button" className={cx({"btn": bootstrapCss,"btn-xs": bootstrapCss})}>Cibler</button></div>
			</div>

			<div className={cx({"row": bootstrapCss})}>
				<div className={cx({"col-md-3": bootstrapCss},"media-icon": bootstrapCss,"text-center": bootstrapCss)}>
					<i className={cx("fa","fa-envelope fa-2x")} aria-hidden="true"></i>
				</div>
				<div className={cx({"col-md-6": bootstrapCss,"media-count": bootstrapCss})}>2</div>
				<div className={cx({"col-md-3": bootstrapCss,"media-target": bootstrapCss})}>
					<button type="button" className={cx({"btn": bootstrapCss,"btn-xs": bootstrapCss})}>Cibler</button></div>
			</div>
			</div>
		);
	}
}

Result.propTypes = {
	numFound: PropTypes.number.isRequired
};

export default Result;
