// solr-faceted-search-react/src/api/solr-query.js
function flatten(arr) {
   return arr.reduce(function (flat, toFlatten) {
     	return flat.concat(Array.isArray(toFlatten) ? flatten(toFlatten) : toFlatten);
   }, []);
}

const rangeFacetToQueryFilter = (field) => {
	const filters = field.value || [];
	if (filters.length < 2) {
		return null;
	}

	return encodeURIComponent(`${field.field}:[${filters[0]} TO ${filters[1]}]`);
  //return `${field.field}:[${filters[0]} TO ${filters[1]}]`;
};

const rangeFacetToJSONQueryFilter = (field) => {
	const filters = field.value || [];
	if (filters.length < 2) {
		return null;
	}

	//return encodeURIComponent(`${field.field}:[${filters[0]} TO ${filters[1]}]`);
  return `${field.field}:[${filters[0]} TO ${filters[1]}]`;
};

const listFacetFieldToQueryFilter = (field) => {
	const filters = field.value || [];
	if (filters.length === 0) {
		return null;
	}

	const filterQ = filters.map((f) => `"${f}"`).join(" OR ");
	return encodeURIComponent(`${field.field}:(${filterQ})`);
};


const listFacetFieldToJSONQueryFilter = (field) => {
	const filters = field.value || [];
	if (filters.length === 0) {
		return null;
	}

	const filterQ = filters.map((f) => `"${f}"`).join(" OR ");
	var filterRender = '';

	if (field.invert) {
		filterRender = `-${field.field}:(${filterQ})`;
	} else {
		filterRender = `${field.field}:(${filterQ})`;
	}

	if (field.multi_select == true) {
		return `{!tag=${field.facet_name}}(${filterRender})`;
	} else {
		//return `${field.field}:(${filterQ})`;}
    if (field.nested == true) {
      return `{!parent which=${field.child}}(${filterRender})`
    } else {
      return `${filterRender}`;
    }
	}
};

const matrixFacetFieldToQueryFilter = (fields) => {
	const filters = field_a.value || [];
	if (filters.length === 0) {
		return null;
	}

	const filterQ = filters.map((f) => `"${f}"`).join(" OR ");
	return encodeURIComponent(`${field.field}:(${filterQ})`);
};

const matrixValueToQueryFilter = (value) => {
	// Turns EDF.CDE.B.-1 into {!parent which=object_type:contact}-(enseigne_s:EDF AND type_s:CDE AND campagne_s:B)
	// Turns INI.KIT.Y.1 into {!parent which=object_type:contact}(enseigne_s:EDF AND type_s:CDE AND campagne_s:B)
	// Turns INI.KIT.Y.0 into an empty string

	const parent_selector = 'object_type_s:contact'
	const line_key = 'enseigne_s';
	const column_key = 'type_cde_s';
	const cell_key = 'campagne_s';
	const kvSeparator = ':';
	const operator = ' AND ';

	const tokens = value.split('.');
	const [line_value,column_value,cell_value,state_value] = tokens;

	const filter = [
		[line_key,line_value].join(kvSeparator),
		[column_key,column_value].join(kvSeparator),
		[cell_key,cell_value].join(kvSeparator)
	].join(operator);

	const query_filter = state_value == 0 ? '' : (
		(state_value < 0 ? '-' : '') +
		`_query_:\"{!parent which=${parent_selector}}(${filter})\"`
		);

	return query_filter;
};

const matrixValueToQueryFilter2 = (value) => {
	// Turns EDF.CDE.B.-1 into {!parent which=object_type:contact}-(enseigne_s:EDF AND type_s:CDE AND campagne_s:B)
	// Turns INI.KIT.Y.1 into {!parent which=object_type:contact}(enseigne_s:EDF AND type_s:CDE AND campagne_s:B)
	// Turns INI.KIT.Y.0 into an empty string

	const parent_selector = 'object_type_s:contact'
	const line_key = 'enseigne_s';
	const column_key = 'type_cde_s';
	const cell_key = 'campagne_s';
	const kvSeparator = ':';
	const operator = ' AND ';

	const tokens = value.split('.');
	const [line_value,column_value,cell_value,state_value] = tokens;

	const filter = [
		[line_key,line_value].join(kvSeparator),
		[column_key,column_value].join(kvSeparator),
		[cell_key,cell_value].join(kvSeparator)
	].join(operator);

	const query_filter = state_value == 0 ? '' : (
		`(${filter})`
		);

	return query_filter;
};

const matrixFacetFieldToJSONQueryFilter = (field) => {
	const filters = field.value || [];
	//if (filters.length === 0) {
	//	return null;
	//}

	const inFilters = filters.filter((v)=> v.split('.')[3] > 0);
	const exFilters = filters.filter((v)=> v.split('.')[3] < 0);
	console.log('------- MATRIX ----- FILTERS -----');
	console.log(inFilters);
	console.log(exFilters);
	console.log('------- MATRIX ----- FILTERS -----');
	const inQuery = inFilters.map((v) => matrixValueToQueryFilter2(v)).filter((v) => v.length > 0);
	const exQuery = exFilters.map((v) => matrixValueToQueryFilter2(v)).filter((v) => v.length > 0);
	var inclusions = '';
	if (inQuery.length > 0) {
		inclusions = ['_query_:\"{!parent which=object_type_s:contact}',inQuery.join(' '),'\"'].join(' ');
	}
	var exclusions = '';
	if (exQuery.length > 0) {
		exclusions = ['-_query_:\"{!parent which=object_type_s:contact}',exQuery.join(' '),'\"'].join(' ');
	}
	return [inclusions,exclusions];
};

const textFieldToQueryFilter = (field) => {
	if(!field.value || field.value.length === 0) {
		return null;
	}

	return encodeURIComponent(field.field === "*" ? field.value : `${field.field}:${field.value}`);
};

const fieldToJSONFacet = (field) => {
	const JSONFacetDef = {};
	JSONFacetDef[field.facet_name] = {}
	JSONFacetDef[field.facet_name].type = 'terms'
	JSONFacetDef[field.facet_name].field = field.field
	field.multi_select ?
		JSONFacetDef[field.facet_name].domain = {excludeTags:field.facet_name}
		: {}
  field.nested ?
    JSONFacetDef[field.facet_name].domain = {blockChildren:field.child}
    : {}
	JSONFacetDef[field.facet_name].sort = field.facetSort
	JSONFacetDef[field.facet_name].limit = -1

	console.debug(JSONFacetDef)
	return JSONFacetDef;
};

const fieldToQueryFilter = (field) => {
	if (field.type === "text") {
		return textFieldToQueryFilter(field);
	} else if (field.type === "list-facet") {
		return listFacetFieldToJSONQueryFilter(field);
	} else if (field.type === "matrix-facet"){
		return matrixFacetFieldToJSONQueryFilter(field);
	} else if (field.type.indexOf("range") > -1) {
		return rangeFacetToJSONQueryFilter(field);
	}
	return null;
};


const buildQuery = (fields) => fields
	.map(fieldToQueryFilter)
	.filter((queryFilter) => queryFilter !== null)
	//.map((queryFilter) => `fq=${queryFilter}`)
	//.join("&");
	//.join(" AND ")
	;

const facetFields = (fields) => fields
	.filter((field) => field.type === "list-facet" || field.type === "range-facet")
	//.map((field) => `facet.field=${encodeURIComponent(field.field)}`)
	//.join("&");
	.map(fieldToJSONFacet);

const facetSorts = (fields) => fields
	.filter((field) => field.facetSort)
	.map((field) => `f.${encodeURIComponent(field.field)}.facet.sort=${field.facetSort}`)
	.join("&");

const buildSort = (sortFields) => sortFields
	.filter((sortField) => sortField.value)
	.map((sortField) => encodeURIComponent(`${sortField.field} ${sortField.value}`))
	.join(",");

const buildFormat = (format) => Object.keys(format)
	.map((key) => `${key}=${encodeURIComponent(format[key])}`)
	.join("&");


const solrQuery2 = (query, format = {wt: "json"}) => { // solr input and output as json
	const {
			searchFields,
			sortFields,
			rows,
			start,
			facetLimit,
			facetSort,
			pageStrategy,
			cursorMark,
			idField
	} = query;

	const filters = (query.filters || []).map((filter) => ({...filter, type: filter.type || "text"}));
	const queryParams = flatten(buildQuery(searchFields.concat(filters)));

	const facetFieldParam = facetFields(searchFields);
	const facetSortParams = facetSorts(searchFields);
	const facetLimitParam = `facet.limit=${facetLimit || -1}`;
	const facetSortParam = `facet.sort=${facetSort || "index"}`;

	const cursorMarkParam = pageStrategy === "cursor" ? `cursorMark=${encodeURIComponent(cursorMark || "*")}` : "";
	const idSort = pageStrategy === "cursor" ? [{field: idField, value: "asc"}] : [];

	const sortParam = buildSort(sortFields.concat(idSort));

	const result = {};

	let facet = {};
	console.debug("Found following facets")
	
	facetFieldParam.map((f) => {
		for (let key in f) facet[key] = f[key];
	});
	console.debug(facet);

	result.query = "object_type_s:contact";
	result.params = {};
	result.params.wt = "json";
	result.filter = queryParams.filter((param) => param.length > 0);
	//result.filter = queryParams.filter((param) => Array.isArray(param) == true);
	result.facet = facet;
	result.sort = sortParam;

	console.debug("Solr query looks like below")
	console.debug(result)

	return JSON.stringify(result, null, 2);

	//return `q=*:*&${queryParams.length > 0 ? queryParams : ""}` +
	//	`${sortParam.length > 0 ? `&sort=${sortParam}` : ""}` +
	//	`${facetFieldParam.length > 0 ? `&${facetFieldParam}` : ""}` +
	//	`${facetSortParams.length > 0 ? `&${facetSortParams}` : ""}` +
	//	`&rows=${rows}` +
	//	`&${facetLimitParam}` +
	//	`&${facetSortParam}` +
	//	`&${cursorMarkParam}` +
	//	(start === null ? "" : `&start=${start}`) +
	//	"&facet=on" +
	//	`&${buildFormat(format)}`;
};

const solrQuery = (query, format = {wt: "json"}) => { //solr input as URI params output as json
	const {
			searchFields,
			sortFields,
			rows,
			start,
			facetLimit,
			facetSort,
			pageStrategy,
			cursorMark,
			idField
		} = query;

	const filters = (query.filters || []).map((filter) => ({...filter, type: filter.type || "text"}));
	const queryParams = buildQuery(searchFields.concat(filters));

	const facetFieldParam = facetFields(searchFields);
	const facetSortParams = facetSorts(searchFields);
	const facetLimitParam = `facet.limit=${facetLimit || -1}`;
	const facetSortParam = `facet.sort=${facetSort || "index"}`;

	const cursorMarkParam = pageStrategy === "cursor" ? `cursorMark=${encodeURIComponent(cursorMark || "*")}` : "";
	const idSort = pageStrategy === "cursor" ? [{field: idField, value: "asc"}] : [];

	const sortParam = buildSort(sortFields.concat(idSort));

	return `q=*:*&${queryParams.length > 0 ? queryParams : ""}` +
			`${sortParam.length > 0 ? `&sort=${sortParam}` : ""}` +
			`${facetFieldParam.length > 0 ? `&${facetFieldParam}` : ""}` +
			`${facetSortParams.length > 0 ? `&${facetSortParams}` : ""}` +
			`&rows=${rows}` +
			`&${facetLimitParam}` +
			`&${facetSortParam}` +
			`&${cursorMarkParam}` +
			(start === null ? "" : `&start=${start}`) +
			"&facet=on" +
			`&${buildFormat(format)}`;
};

export default solrQuery2;


export {
	rangeFacetToQueryFilter,
	listFacetFieldToQueryFilter,
	textFieldToQueryFilter,
	fieldToQueryFilter,
	buildQuery,
	facetFields,
	facetSorts,
	buildSort,
	solrQuery,
	solrQuery2
};
